pandas/core/internals.py

```diff
@@ -1947,20 +1947,42 @@ def _try_fill(self, value):
     def fillna(self, value, limit=None,
                inplace=False, downcast=None):
 
-        # straight putmask here
-        values = self.values if inplace else self.values.copy()
         mask = isnull(self.values)
         value = self._try_fill(value)
+
         if limit is not None:
             if self.ndim > 2:
                 raise NotImplementedError("number of dimensions for 'fillna' "
                                           "is currently limited to 2")
             mask[mask.cumsum(self.ndim-1)>limit]=False
 
-        np.putmask(values, mask, value)
-        return [self if inplace else
-                self.make_block(values,
-                                fastpath=True)]
+        if mask.any():
+            try:
+                return self._fillna_mask(mask, value, inplace=inplace)
+            except TypeError:
```

Comment:
I added ``mask.any()`` condition to handle the case when the block contains no ``Nan``. In this case:

- We can return the block as it is. 
- Don't want to coerce data to ``object`` even if the fill value is ``object``.
- ``DatetimeTZBlock`` can perform inplace op in this case. Otherwise can't because it's internal value is ``DatetimeIndex``.

Let me do a follow-up if I do something wrong.